text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1006 @@
+
+[motivation]: #motivation
+
+## Arbitrary tuple arity support
+
+Currently, when a user wants to either use or add behavior to tuples, he writes an impl for each tuple arity.
+For easier maintenance, it is usually done with a `macro_rules` and implemented up to 12 arity tuple. (see [ `Hash` implementation in `std`](https://github.com/rust-lang/rust/blob/master/src/libcore/hash/mod.rs)).
+
+Variadic tuple will provide several benefits considering trait implementation for tuple or using variadic tuples:
+
+- Implementations will be easier to write
+- Implementations will be easier to read and maintain
+- The compiler will compile implementation only for required tuple arity
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Let's call a _variadic tuple type_ a tuple type with an arbitrary arity and a _variadic tuple_ an instance of a variadic tuple type.
+
+A variadic tuple type is declared with `(..T)` and a variadic tuple type can be expanded with `(..Vec<T>)`.
+
+Note: To illustrate the RFC, we will use the current implementation of the `Hash` trait for tuples.
+
+```rust
+// Quote from Rust source code
+// This macro implements `Hash` for a tuple.
+// It is used like this: `impl_hash_tuple! { A B C D E F }` for a 6-arity tuple.
+macro_rules! impl_hash_tuple {
+    () => (
+        #[stable(feature = "rust1", since = "1.0.0")]
+        impl Hash for () {
+            fn hash<H: Hasher>(&self, _state: &mut H) {}
+        }
+    );
+
+    ( $($name:ident)+) => (
+        #[stable(feature = "rust1", since = "1.0.0")]
+        impl<$($name: Hash),+> Hash for ($($name,)+) where last_type!($($name,)+): ?Sized {
+            #[allow(non_snake_case)]
+            fn hash<S: Hasher>(&self, state: &mut S) {
+                let ($(ref $name,)+) = *self;
+                $($name.hash(state);)+
+            }
+        }
+    );
+}
+
+macro_rules! last_type {
+    ($a:ident,) => { $a };
+    ($a:ident, $($rest_a:ident,)+) => { last_type!($($rest_a,)+) };
+}
+```
+
+## Variadic tuple type
+
+### Declaration
+
+Variadic tuple types are always declared in a generic parameter group.
+
+There are two different syntaxes:
+
+1. `(..T)`: declare a single variadic tuple type identified by `T`
+2. `(..(T1, T2, ..., Tn))`: declare n variadic tuple types identified by `T1`, `T2`, ..., `Tn`, all these variadic tuple types have the same arity.
+
+Declaration examples:
+
+- `struct VariadicStruct<(..T1)>` : declares a struct with a variadic tuple type identified by `T1` in its generic parameters
+- `fn my_func<(..T1), (..T2)>()`: a function can have variadic tuple type parameters
+- `impl<(..Head)>`: is an implementation block that uses a variadic tuple type identified by `Head`
+- `impl<A, B, C, (.._Tail)>`:  same as above, but with other generic parameters
+- `impl<A, B, (..C), (..D)>`: there can be several variadic tuple types declared in a generic parameter group
+- `impl<A, B, (..(C, D)), (..E)>`: and we can mix both syntaxes
```

Comment:
Having some really concrete use cases would help. Names like my_func and VariadicStruct really don't help.

Honestly, the places where you can use these isn't really the thing that needs examples; I'd rather have examples of variadic tuples that mix the syntax. So, replace the list with things like `(..T)`, `(Head, ..Tail)`, `(First, ..Middle, End)`, etc.

Reply:
This section is about declarations of variadic type parameters.  IIUC, the proposal does not permit `(Head, ..Tail)` in a declaration.

Reply:
Wait, there are variadic type parameters options? I thought this was just for variadic tuple types.